core/api/utils.hpp

```diff
@@ -27,6 +21,67 @@ namespace fc::api {
   using ParamsTuple =
       std::tuple<std::remove_const_t<std::remove_reference_t<T>>...>;
 
+  template <typename T, typename... Ts>
+  class ApiMethod {
+   public:
+    using Result = T;
+    using OutcomeResult = outcome::result<T>;
+    using Params = ParamsTuple<Ts...>;
+    using Callback = std::function<void(OutcomeResult)>;
+    using FunctionSimpleSignature = OutcomeResult(Ts...);
+    using FunctionSignature = void(Callback, Ts...);
+
+    OutcomeResult operator()(Ts... args) const {
+      if (!f_) {
+        return ERROR_TEXT("API not set up");
+      }
+      std::promise<OutcomeResult> wait;
+      f_([&wait](const OutcomeResult &res) { return wait.set_value(res); },
+         args...);
+      return wait.get_future().get();
+    }
+    outcome::result<void> operator()(
+        const std::function<void(OutcomeResult)> &cb, Ts... args) const {
```

Comment:
```suggestion
    void operator()(
        Callback cb, Ts... args) const {
```
async functions should return void, and pass errors to callback.

`std::function` already throws [`std::bad_function_call`](https://en.cppreference.com/w/cpp/utility/functional/function/operator()#Exceptions).

also, that errror must not trigger on production, because api setup is hardcoded and expected to be correct. such error should be detected during testing and debugging.